service/matching/matchingEngine.go

```diff
@@ -860,23 +837,42 @@ func (e *matchingEngineImpl) GetTaskQueueUserData(
 	if version < 0 {
 		return nil, serviceerror.NewInvalidArgument("last_known_user_data_version must not be negative")
 	}
-	resp := &matchingservice.GetTaskQueueUserDataResponse{}
-	userData, err := tqMgr.GetUserData(ctx)
-	if err != nil {
-		return nil, err
+
+	if req.WaitNewData {
+		var cancel context.CancelFunc
+		ctx, cancel = newChildContext(ctx, e.config.GetUserDataLongPollTimeout(), returnEmptyTaskTimeBudget)
```

Comment:
Seems like this should be a caller concern, why set the deadline in the handler?

Reply:
Look at what newChildContext does: it's trimming off a second at the end to leave time to return a result

Reply:
Got it.